File: backend/app/project/api/models/configuration.py
# SPDX-FileCopyrightText: 2020 - 2023
# - Martin Abbrent <martin.abbrent@ufz.de>
# - Kotyba Alhaj Taha <kotyba.alhaj-taha@ufz.de>
# - Nils Brinckmann <nils.brinckmann@gfz-potsdam.de>
# - Marc Hanisch <marc.hanisch@gfz-potsdam.de>
# - Helmholtz Centre Potsdam - GFZ German Research Centre for Geosciences (GFZ, https://www.gfz-potsdam.de)
# - Helmholtz Centre for Environmental Research GmbH - UFZ (UFZ, https://www.ufz.de)
#
# SPDX-License-Identifier: HEESIL-1.0

"""Class and helpers for the configurations."""

from ..es_utils import ElasticSearchIndexTypes, settings_with_ngrams
from ..helpers.errors import ConflictError
from .base_model import db
from .mixin import (
    ArchivableMixin,
    AuditMixin,
    BeforeCommitValidatableMixin,
    SearchableMixin,
)


class Configuration(
    db.Model, AuditMixin, ArchivableMixin, SearchableMixin, BeforeCommitValidatableMixin
):
    """Data model for the configurations."""

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    start_date = db.Column(db.DateTime(timezone=True), nullable=True)
    end_date = db.Column(db.DateTime(timezone=True), nullable=True)
    label = db.Column(db.String(256), nullable=True)
    project = db.Column(db.String(256), nullable=True)
    description = db.Column(db.Text, nullable=True)
    status = db.Column(db.String(256), nullable=True, default="draft")
    cfg_permission_group = db.Column(db.String, nullable=True)
    is_internal = db.Column(db.Boolean, default=False)
    is_public = db.Column(db.Boolean, default=False)
    update_description = db.Column(db.String(256), nullable=True)
    b2inst_record_id = db.Column(db.String(256), nullable=True)
    persistent_identifier = db.Column(db.String(256), nullable=True, unique=True)
    site_id = db.Column(db.Integer, db.ForeignKey("site.id"), nullable=True)
    site = db.relationship("Site", backref="configurations")

    def validate(self):
        """
        Validate the model.

        Check that we don't have multiple visibility values.
        """
        super().validate()
        if self.is_internal and self.is_public:
            raise ConflictError("The configuration can't be both internal and public")

    def to_search_entry(self):
        """
        Return the configuration as dict for full text search.

        All the fields here will be searchable and can be used as
        filters in our full text search.
        """
        return {
            "label": self.label,
            "description": self.description,
            "project": self.project,
            "status": self.status,
            "cfg_permission_group": self.cfg_permission_group,
            "updated_at": self.updated_at,
            "configuration_contact_roles": [
                ccr.to_search_entry() for ccr in self.configuration_contact_roles
            ],
            "attachments": [
                a.to_search_entry() for a in self.configuration_attachments
            ],
            "generic_actions": [
                g.to_search_entry() for g in self.generic_configuration_actions
            ],
            "configuration_static_location_actions": [
                s.to_search_entry()
                for s in self.configuration_static_location_begin_actions
            ],
            "configuration_dynamic_location_actions": [
                d.to_search_entry()
                for d in self.configuration_dynamic_location_begin_actions
            ],
            "platform_mount_actions": [
                p.to_search_entry() for p in self.platform_mount_actions
            ],
            "device_mount_actions": [
                d.to_search_entry() for d in self.device_mount_actions
            ],
            "configuration_customfields": [
                a.to_search_entry() for a in self.configuration_customfields
            ],
            "configuration_parameters": [
                p.to_search_entry() for p in self.configuration_parameters
            ],
            "archived": self.archived,
            "is_internal": self.is_internal,
            "is_public": self.is_public,
            "created_by_id": self.created_by_id,
            "start_date": self.start_date,
            "end_date": self.end_date,
            "site_id": self.site_id,
            "persistent_identifier": self.persistent_identifier,
        }

    @staticmethod
    def get_search_index_definition():
        """
        Return the index configuration for the elasticsearch.

        Describes which fields will be searchable by some text (with stemmer, etc)
        and via keyword (raw equality checks).
        """
        from ..models.contact import Contact
        from ..models.device import Device
        from ..models.platform import Platform

        type_keyword = ElasticSearchIndexTypes.keyword()
        type_text_full_searchable = ElasticSearchIndexTypes.text_full_searchable(
            analyzer="ngram_analyzer"
        )
        type_keyword_and_full_searchable = (
            ElasticSearchIndexTypes.keyword_and_full_searchable(
                analyzer="ngram_analyzer"
            )
        )

        return {
            "aliases": {},
            "mappings": {
                "properties": {
                    "is_internal": {
                        "type": "boolean",
                    },
                    "is_public": {
                        "type": "boolean",
                    },
                    "archived": {
                        "type": "boolean",
                    },
                    "created_by_id": {
                        "type": "integer",
                    },
                    "label": type_keyword_and_full_searchable,
                    "project": type_keyword_and_full_searchable,
                    "description": type_text_full_searchable,
                    "status": type_keyword_and_full_searchable,
                    "cfg_permission_group": type_keyword,
                    "start_date": {"type": "date"},
                    "end_date": {"type": "date"},
                    "updated_at": {
                        "type": "date",
                        "format": "strict_date_optional_time",
                    },
                    "platforms": {
                        "properties": Platform.get_search_index_properties(),
                    },
                    "devices": {
                        "properties": Device.get_search_index_properties(),
                    },
                    "configuration_contact_roles": {
                        "properties": {
                            "role_name": type_keyword_and_full_searchable,
                            "role_uri": type_keyword,
                            "contact": {
                                "properties": Contact.get_search_index_properties(),
                            },
                        },
                    },
                    "attachments": {
                        "properties": {
                            # Allow search via text & keyword
                            "label": type_keyword_and_full_searchable,
                            # But don't allow search for the very same url (unlikely to be needed).
                            "url": type_text_full_searchable,
                        },
                    },
                    "configuration_parameters": {
                        "properties": {
                            "label": type_keyword_and_full_searchable,
                            "description": type_keyword_and_full_searchable,
                            "unit_uri": type_keyword,
                            "unit_name": type_keyword_and_full_searchable,
                            "configuration_parameter_value_change_actions": {
                                "properties": {
                                    "value": type_keyword_and_full_searchable,
                                    "description": type_keyword_and_full_searchable,
                                },
                            },
                        },
                    },
                    "generic_actions": {
                        "properties": {
                            "action_type_uri": type_keyword,
                            "action_type_name": type_keyword_and_full_searchable,
                            "description": type_text_full_searchable,
                        },
                    },
                    "configuration_static_location_actions": {
                        "properties": {
                            "begin_description": type_text_full_searchable,
                            "end_description": type_text_full_searchable,
                            "label": type_keyword_and_full_searchable,
                        },
                    },
                    "configuration_dynamic_location_actions": {
                        "properties": {
                            "begin_description": type_text_full_searchable,
                            "end_description": type_text_full_searchable,
                            "label": type_keyword_and_full_searchable,
                        },
                    },
                    "platform_mount_actions": {
                        "properties": {
                            "begin_description": type_text_full_searchable,
                            "end_description": type_text_full_searchable,
                            "platform": {
                                "properties": Platform.get_search_index_properties(),
                            },
                        },
                    },
                    "device_mount_actions": {
                        "properties": {
                            "begin_description": type_text_full_searchable,
                            "end_description": type_text_full_searchable,
                            "device": {
                                "properties": Device.get_search_index_properties(),
                            },
                        },
                    },
                    "customfields": {
                        "properties": {
                            # The key should use keyword behaviour by default
                            # but should also searchable as text.
                            "key": type_keyword_and_full_searchable,
                            # The same for the value.
                            "value": type_keyword_and_full_searchable,
                        },
                    },
                    "site_id": {
                        "type": "integer",
                    },
                    "persistent_identifier": type_keyword_and_full_searchable,
                }
            },
            "settings": settings_with_ngrams(
                analyzer_name="ngram_analyzer",
                filter_name="ngram_filter",
                min_ngram=1,
                max_ngram=10,
                max_ngram_diff=10,
            ),
        }

File: frontend/models/Configuration.ts
/**
 * @license EUPL-1.2
 * SPDX-FileCopyrightText: 2020 - 2024
 * - Nils Brinckmann <nils.brinckmann@gfz-potsdam.de>
 * - Marc Hanisch <marc.hanisch@gfz-potsdam.de>
 * - Helmholtz Centre Potsdam - GFZ German Research Centre for Geosciences (GFZ, https://www.gfz-potsdam.de)
 *
 * SPDX-License-Identifier: EUPL-1.2
 */
import { DateTime } from 'luxon'

import { Image, IImage } from '@/models/Image'
import { IContact, Contact } from '@/models/Contact'
import { IMountActions } from '@/models/IMountActions'
import { DeviceMountAction } from '@/models/DeviceMountAction'
import { PlatformMountAction } from '@/models/PlatformMountAction'
import { PermissionGroup, IPermissionGroup, IPermissionableSingleGroup } from '@/models/PermissionGroup'
import { Visibility, IVisible } from '@/models/Visibility'
import { Parameter, IParameter } from '@/models/Parameter'

export interface IConfiguration extends IMountActions, IPermissionableSingleGroup {
  id: string
  persistentIdentifier: string
  startDate: DateTime | null
  endDate: DateTime | null
  label: string
  description: string
  project: string
  campaign: string
  status: string
  archived: boolean
  contacts: IContact[]
  parameters: IParameter[]
  images: IImage[]
  createdAt: DateTime | null
  updatedAt: DateTime | null
  updateDescription: string
  createdBy: IContact | null
  updatedBy: IContact | null
  /*
    You may wonder why there is an extra createdByUserId entry here.
    The reason is that we use the createBy & updatedBy to show
    information about the contact that are responsible for the changes.
    Here we refer to the user object (which can have different ids).
  */
  createdByUserId: string | null
  visibility: Visibility
  siteId?: string
  keywords: string[]
}

export class Configuration implements IConfiguration, IVisible {
  private _id: string = ''
  private _persistentIdentifier: string = ''
  private _startDate: DateTime | null = null
  private _endDate: DateTime | null = null
  private _label: string = ''
  private _description: string = ''
  private _project: string = ''
  private _campaign: string = ''
  private _status: string = ''
  private _archived: boolean = false
  private _contacts: IContact[] = [] as IContact[]
  private _parameters: Parameter[] = []
  private _images: Image[] = []
  private _deviceMountActions: DeviceMountAction[] = []
  private _platformMountActions: PlatformMountAction[] = []
  private _permissionGroup: IPermissionGroup | null = null
  private _createdAt: DateTime | null = null
  private _updatedAt: DateTime | null = null
  private _updateDescription: string = ''
  private _createdBy: IContact | null = null
  private _updatedBy: IContact | null = null
  private _createdByUserId: string | null = null
  private _visibility: Visibility = Visibility.Internal
  private _siteId: string = ''
  private _keywords: string[] = []

  get id (): string {
    return this._id
  }

  set id (id: string) {
    this._id = id
  }

  get persistentIdentifier (): string {
    return this._persistentIdentifier
  }

  set persistentIdentifier (persistentIdentifier: string) {
    this._persistentIdentifier = persistentIdentifier
  }

  get startDate (): DateTime | null {
    return this._startDate
  }

  set startDate (date: DateTime | null) {
    this._startDate = date
  }

  get endDate (): DateTime | null {
    return this._endDate
  }

  set endDate (date: DateTime | null) {
    this._endDate = date
  }

  get label (): string {
    return this._label
  }

  set label (newLabel: string) {
    this._label = newLabel
  }

  get description (): string {
    return this._description
  }

  set description (newDescription: string) {
    this._description = newDescription
  }

  get project (): string {
    return this._project
  }

  set project (newProject: string) {
    this._project = newProject
  }

  get campaign (): string {
    return this._campaign
  }

  set campaign (newCampaign: string) {
    this._campaign = newCampaign
  }

  get status (): string {
    return this._status
  }

  set status (newStatus: string) {
    this._status = newStatus
  }

  get contacts (): IContact[] {
    return this._contacts
  }

  set contacts (contacts: IContact[]) {
    this._contacts = contacts
  }

  get deviceMountActions (): DeviceMountAction[] {
    return this._deviceMountActions
  }

  set deviceMountActions (newDeviceMountActions: DeviceMountAction[]) {
    this._deviceMountActions = newDeviceMountActions
  }

  get platformMountActions (): PlatformMountAction[] {
    return this._platformMountActions
  }

  set platformMountActions (newPlatformMountActions: PlatformMountAction[]) {
    this._platformMountActions = newPlatformMountActions
  }

  get parameters (): Parameter[] {
    return this._parameters
  }

  set images (images: Image[]) {
    this._images = images
  }

  get images (): Image[] {
    return this._images
  }

  set parameters (parameters: Parameter[]) {
    this._parameters = parameters
  }

  get permissionGroup (): IPermissionGroup | null {
    return this._permissionGroup
  }

  set permissionGroup (permissionGroup: IPermissionGroup | null) {
    this._permissionGroup = permissionGroup
  }

  get createdAt (): DateTime | null {
    return this._createdAt
  }

  set createdAt (newCreatedAt: DateTime | null) {
    this._createdAt = newCreatedAt
  }

  get updatedAt (): DateTime | null {
    return this._updatedAt
  }

  set updatedAt (newUpdatedAt: DateTime | null) {
    this._updatedAt = newUpdatedAt
  }

  get updateDescription (): string {
    return this._updateDescription
  }

  set updateDescription (newDescription: string) {
    this._updateDescription = newDescription
  }

  get createdBy (): IContact | null {
    return this._createdBy
  }

  set createdBy (user: IContact | null) {
    this._createdBy = user
  }

  get updatedBy (): IContact | null {
    return this._updatedBy
  }

  set updatedBy (user: IContact | null) {
    this._updatedBy = user
  }

  get createdByUserId (): string | null {
    return this._createdByUserId
  }

  set createdByUserId (newId: string | null) {
    this._createdByUserId = newId
  }

  get visibility (): Visibility {
    return this._visibility
  }

  set visibility (visibility: Visibility) {
    this._visibility = visibility
  }

  get isInternal (): boolean {
    return this._visibility === Visibility.Internal
  }

  get isPublic (): boolean {
    return this._visibility === Visibility.Public
  }

  get archived (): boolean {
    return this._archived
  }

  set archived (newValue: boolean) {
    this._archived = newValue
  }

  get type (): string {
    return 'configuration'
  }

  get siteId (): string {
    return this._siteId
  }

  set siteId (id: string) {
    this._siteId = id
  }

  get keywords (): string[] {
    return this._keywords
  }

  set keywords (newKeywords: string[]) {
    this._keywords = newKeywords
  }

  static createFromObject (someObject: IConfiguration): Configuration {
    const newObject = new Configuration()

    newObject.id = someObject.id
    newObject.persistentIdentifier = someObject.persistentIdentifier
    // luxon DateTime objects are immutable
    newObject.startDate = someObject.startDate
    newObject.endDate = someObject.endDate

    newObject.label = someObject.label
    newObject.description = someObject.description
    newObject.project = someObject.project
    newObject.campaign = someObject.campaign
    newObject.status = someObject.status

    newObject.contacts = someObject.contacts.map(Contact.createFromObject)
    newObject.deviceMountActions = someObject.deviceMountActions.map(DeviceMountAction.createFromObject)
    newObject.platformMountActions = someObject.platformMountActions.map(PlatformMountAction.createFromObject)

    newObject.parameters = someObject.parameters.map(Parameter.createFromObject)
    newObject.images = someObject.images.map(Image.createFromObject)
    newObject.permissionGroup = someObject.permissionGroup ? PermissionGroup.createFromObject(someObject.permissionGroup) : null

    newObject.createdAt = someObject.createdAt
    newObject.updatedAt = someObject.updatedAt
    newObject.updateDescription = someObject.updateDescription
    newObject.createdBy = someObject.createdBy ? Contact.createFromObject(someObject.createdBy) : null
    newObject.updatedBy = someObject.updatedBy ? Contact.createFromObject(someObject.updatedBy) : null
    newObject.createdByUserId = someObject.createdByUserId

    newObject.visibility = someObject.visibility
    newObject.archived = someObject.archived

    newObject.siteId = someObject.siteId ?? ''
    newObject.keywords = [...someObject.keywords]

    return newObject
  }
}
